src/scheduler/pbs_sched.cpp

```diff
@@ -530,204 +506,117 @@ close_server(sched_svrconn *sconn)
+	int i;
+
+	pbs_disconnect(clust_primary_sock);
+	pbs_disconnect(clust_secondary_sock);
+
 	if (poll_context != NULL) {
 		tpp_em_destroy(poll_context);
 		poll_context = NULL;
 	}
-	while (!ds_queue_is_empty(sched_cmds)) {
-		sched_cmd *cmd = static_cast<sched_cmd *>(ds_dequeue(sched_cmds));
-		free_sched_cmd(cmd);
-	}
-	free_ds_queue(sched_cmds);
-	sched_cmds = NULL;
-}
 
-/**
- * @brief connect to given host and send register sched request on it
- *        and wait for server reply
- *
- * @param[in] - host       - address of server in host[:port] format
- * @param[in] - is_primary - is primary connection or secondary connection
- *                           0 - secondary else primary
- *
- * @return int
- * @retval -1  - failure
- * @return !-1 - success, and opened connection sock
- */
-static int
-connect_server_helper(const char *host, int is_primary)
-{
-	int rc;
-	int sock = -1;
-	struct batch_reply *reply = NULL;
-
-	sock = pbs_connect(const_cast<char *>(host));
-	if (sock < 0)
-		goto rerr;
-	rc = encode_DIS_ReqHdr(sock, PBS_BATCH_RegisterSched, pbs_current_user);
-	if (rc != DIS_SUCCESS)
-		goto rerr;
-	rc = diswst(sock, sc_name);
-	if (rc != DIS_SUCCESS)
-		goto rerr;
-	rc = encode_DIS_ReqExtend(sock, NULL);
-	if (rc != DIS_SUCCESS)
-		goto rerr;
-	if (dis_flush(sock) != 0)
-		goto rerr;
-	pbs_errno = 0;
-	reply = PBSD_rdrpy(sock);
-	if (reply == NULL)
-		goto rerr;
-	if (pbs_errno != 0) {
-		char *errmsg = get_conn_errtxt(sock);
-		if (errmsg) {
-			log_eventf(PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_NOTICE,
-				   msg_daemonname, "Server rejected register request for %s connection with error: %s",
-				   is_primary ? "primary" : "secondary", errmsg);
-			die(-1);
-		}
-		goto rerr;
+	/* free qrun_list */
+	for (i = 0; i < qrun_list_size; i++) {
+		if (qrun_list[i].jid != NULL)
+			free(qrun_list[i].jid);
 	}
+	free(qrun_list);
 
-	PBSD_FreeReply(reply);
-	return sock;
+	sched_svr_init();
 
-rerr:
-	pbs_disconnect(sock);
-	PBSD_FreeReply(reply);
-	return -1;
+	clust_primary_sock = -1;
+	clust_secondary_sock = -1;
 }
 
 /**
- * @brief connect to given server, create and add server struct in servers list
- *        also add secondary connection to poll list
+ * @brief connect to all the servers configured. Also add secondary connections
+ *              of individual servers secondary sd's to the poll list
  *
- * @param[in] svrhost - address of server in host[:port] format
  *
  * @return void
  */
 static void
-connect_server(const char *svrhost)
+connect_svrpool()
 {
-	int primary_sock = -1;
-	int secondary_sock = -1;
-	sched_svrconn *svr;
-	void *svrs_newarr;
-
-	if (svrhost == NULL)
-		return;
-
-	while (primary_sock < 0 || secondary_sock < 0) {
-		if (primary_sock < 0) {
-			primary_sock = connect_server_helper(svrhost, 1);
-			if (primary_sock < 0) {
+	int i;
+	svr_conn_t *svr_conns;
+	int registered = 0;
+
+
+	while (clust_primary_sock < 0 || clust_secondary_sock < 0 || registered == 0) {
```

Comment:
Earlier the function was connecting to only one server, now the function is meant to connect to a server pool. But this while loop does not really loop on the server pool. 
Since pbs_connect() connects to all servers but returns only one descriptor, there is no way you can tell that it connected to all server instances or not.

Reply:
As discussed pbs_connect() connects to all servers and returns the first socket which later can be used to fetch all the servers in that set.
Also just added couple of checks to make sure whether pbs_connect indeed connected to all the configured servers or not. If not connected then we retry pbs_connect.